### ISSUE TITLE
CA2200: Rethrow to preserve stack details - should not warn when throwing different exception

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/roslyn-analyzers#4280]
### Analyzer

**Diagnostic ID**: [CA2200](https://docs.microsoft.com/visualstudio/code-quality/ca2200): `Rethrow to preserve stack details`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100-preview.8](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug
The caught exception `ex` is only thrown when it is mapped into a different exception by the `MapSendException(...)` method. The analyzer should not warn in this case, further I would assume CA2200 should not warn when the catch block is throwing a different exception (`mappedEx`) than the catched exception (`ex`)

```csharp
    try
    {
          ... 
    }
    catch (Exception ex) when (MapSendException(ex, cancellationToken, out Exception mappedEx))
    {
        throw mappedEx;
    }
```

The exact warning found [here](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L751), related to https://github.com/dotnet/runtime/pull/42957#discussion_r498532030

### Expected behavior
Should not warn in above case
### Actual behavior
Warning

**Comment by Youssef1313** [comment]
A similar case is being tested:

https://github.com/dotnet/roslyn-analyzers/blob/76374e785b11c2fab3515cf35f89f5f2c4c0eeeb/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetailsTests.cs#L56

@Evangelink Was this fixed in your re-write of the analyzer? or it should have been working from before the rewrite?

**Comment by mavasani** [comment]
@buyaa-n Can you test with the latest package from https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet6&package=Microsoft.CodeAnalysis.NetAnalyzers&protocolType=NuGet&version=6.0.0-preview1.20501.5&view=overview? I suspect @Evangelink's recent fix in master would have fixed this. We should consider backporting the change to release/dotnet5-rc2 branch if that is the case.

**Comment by buyaa-n** [comment]
> A similar case is being tested:
> @Evangelink Was this fixed in your re-write of the analyzer? 

Yep, it is the same case, looks already fixed

>  @buyaa-n Can you test with the latest package from https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet6&package=Microsoft.CodeAnalysis.NetAnalyzers&protocolType=NuGet&version=6.0.0-preview1.20501.5&view=overview? I suspect @Evangelink's recent fix in master would have fixed this. We should consider backporting the change to release/dotnet5-rc2 branch if that is the case.

Yep i used RC2 branch, not master, agree we should port it to RC2

**Comment by buyaa-n** [comment]
> @buyaa-n Can you test with the latest package from https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet6&package=Microsoft.CodeAnalysis.NetAnalyzers&protocolType=NuGet&version=6.0.0-preview1.20501.5&view=overview

Hm, i still see the warning with the that version

**Comment by Youssef1313** [comment]
@buyaa-n  Interesting! Let me add a test and debug it.

**Comment by Evangelink** [comment]
I didn't consider this case, honestly that's the first time I am seeing this pattern :)

@Youssef1313 already has filed a PR
